### PR TITLE
[ADD] convert_from_company_dependent

### PR DESCRIPTION
**Pull request by hbrunn**
this is the function I propose in https://github.com/OCA/OpenUpgrade/pull/4457#discussion_r1819214161, will be useful elsewhere too

### CODE REVIEW

**Comment by pedrobaeza** [comment]
But I think this depends on the business logic behind, as you are converting a multi-valued field (potentially one per company) into only one value, so you need to decide what to do in case of conflict depending on the use case.

**Comment by hbrunn** [comment]
I think the general case will be pretty much the same everywhere:

v(n-1): model has no company_id and property property_whatever
v(n): model has company_id, property_whatever becomes field whatever

Then the user of the function duplicates every row of model for each company where it makes sense (that's the business logic depending part), and calls `convert_from_company_dependent(model, property_whatever, whatever)`, which is invariant wrt business logic

**Comment by pedrobaeza** [comment]
But in such duplication, `res_id` will be another, and thus, this method won't serve for transferring the value of the corresponding property for the duplicated records.

**Comment by hbrunn** [comment]
it will, because the caller sets company_id for the cloned records as it's appropriate.

consider the [account migration](https://github.com/OCA/OpenUpgrade/pull/4457/files#diff-406c7e73ba75feeb54738e35feff14fb8fe419b838b4461e864cf033b2fa0f52R480), there they duplicate tax groups as it makes sense into the different companies, and in the next function I can replace the horrible incomplete sql statements with three calls to the function I propose.

Wherever Odoo decides to go from properties to company aware on model level, we need the exact same thing.

File: openupgradelib/openupgrade.py [hunk 1]
```diff
@@ -3458,6 +3466,75 @@ def convert_to_company_dependent(
...
+    return logged_query(
+        env.cr,
+        f"""
+        UPDATE {table_name}
+        SET {destination_field_name} = (
+            SELECT {value_expression}
+            FROM ir_property ip
+            WHERE ip.fields_id={field_id} --- {origin_field_name}
+            AND (
+                ip.res_id = '{model_name}.' || {table_name}.id
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade.py:3524]: I insist that this is the problem. The duplicated record has an ID that is not on `ir_property` table, as the corresponding one is the previous record ID.

**Comment by hbrunn** [comment]
aaah wait, I get it now. This indeed needs the extra field I propose on the PR to link the clones to their origin record. That's a good thing to have in those cases anyways, will do next week

File: openupgradelib/openupgrade.py (hunk 1 above)
Comment (hbrunn) [reply openupgradelib/openupgrade.py:3524]: thanks for insisting, see below

**Comment by hbrunn** [comment]
I've just grepped through the other analysis files and didn't find another place where we need this for the v17 migration, so for the time being I put the function in the v17 account migration, and we move it to openupgradelib when the next similar case comes up

**Pull request closed without merging**